checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -334,4 +426,147 @@ private static String regexErrorMessage(String s, int expectedGroups, int actual
+
+        // Special construct means either pure, non-capturing groups that do not capture text and do
+        // not count towards the group total.
+
+        final int length = regexp.length();
+        for (int i = 0; i < length; i++) {
+            // Marks the opening of either a capturing a capturing group or some special
+            // construct.
+            if (regexp.charAt(i) == '(') {
+                boolean isThisCapturingGroup = false;
+                // If '(' is not followed by a '?' or is followed by '?<', it is a capturing group.
+                if ((i < length - 1)
+                        && (regexp.charAt(i + 1) != '?' || regexp.startsWith("?<", i + 1))) {
+                    group += 1;
+                    // Push the group index onto a stack.
+                    unclosedCapturingGroups.push(group);
+                    // This opening bracket marked the opening of a capturing group.
+                    isThisCapturingGroup = true;
+                }
+                // Push the boolean on top of a stack. When we encounter a ')', we will use this
+                // stack to check whether it closes a capturing group or a special construct.
+                isCapturingGroup.push(isThisCapturingGroup);
+                if (isThisCapturingGroup) {
+                    // If this opening was preceded by a '|', this group is optional.
+                    if (i > 0 && regexp.charAt(i - 1) == '|') {
+                        nonNullGroups.remove(Integer.valueOf(group));
+                    }
+                }
+            } else if (regexp.charAt(i) == ')') {
+                // This closes the last opened "thing" (either a capturing group or a special
+                // construct). What the last opened "thing" was, is determined from the boolean
+                // stack.
+                // If the top is false, then it was a special construct.
+                // If the top is true, then it was a capturing group and we need to analyse it
+                // further.
+                boolean closesCapturingGroup = isCapturingGroup.pop();
+                if (closesCapturingGroup) {
+                    // The capturing group being closed is on top of the unclosedCapturingGroups
+                    // stack.
+                    Integer closedGroupIndex = unclosedCapturingGroups.pop();
+                    // If the ')' is followed by a '?', '*', '|' or '{0', then this capturing group
+                    // is optional and can be removed from the list of nonNullGroups.
+                    if ((i < length - 1 && "?*|".contains(String.valueOf(regexp.charAt(i + 1))))
+                            || (i < length - 2 && regexp.startsWith("{0", i + 1))) {
+                        nonNullGroups.remove(closedGroupIndex);
+                    }
+                }
+            } else if (regexp.charAt(i) == '[') {
+                // A character class has been opened. We will traverse till we close this class.
```

Comment:
I suggest abstracting out this body into a separate method, much like `getLastIndexOfEscapeConstruct`.